src/hotspot/share/classfile/systemDictionaryShared.cpp

```diff
@@ -515,8 +517,12 @@ void SystemDictionaryShared::init_dumptime_info(InstanceKlass* k) {
 
 void SystemDictionaryShared::remove_dumptime_info(InstanceKlass* k) {
   MutexLocker ml(DumpTimeTable_lock, Mutex::_no_safepoint_check_flag);
-  get_info_locked(k)->free_saved_cpcache_entries();
   _dumptime_table->remove(k);
+
+  ConstantPoolCache* cpc = k->constants()->cache();
+  if (cpc != NULL) {
+    remove_saved_cpcache_entries_locked(cpc);
+  }
```

Comment:
There are two reasons:

- The old table was indexed by `InstanceKlass*`. However, with class redefinition, a second ConstantPoolCache may be created for a class. Trying to keep this one-to-many relationship with the old table is doable but awkward.
- The old table was cloned when the archive is dumped. However, the saved_cpcache_entries don't need to be cloned because they are never modified. I don't want to have to explain why parts of the DumpTimeClassInfo are not deep-copied.